src/Storages/StorageAzureBlob.cpp

```diff
@@ -882,6 +882,7 @@ StorageAzureBlobSource::Iterator::Iterator(
 
 RelativePathWithMetadata StorageAzureBlobSource::Iterator::next()
 {
+    std::lock_guard lock(next_mutex);
```

Comment:
I think we need to lock mutex only in `else` branch. In case of `keys` mutex is not needed, atomic `index` is enough

Reply:
> Seems like is_finished is also needed only in else branch.
ya indeed, the if part can be written without is_finished, but it helps avoid multiple checks every time and uniformly indicates the state of the iterator.